### PR TITLE
Switch to the Dune build system and ppxlib

### PR DESCRIPTION
**Pull request by MisterDA**
Hi!
This PR switches extunix to the Dune build system, and updates the code onto OCaml 4.11.
Dune prefers libraries and executable to live in different directories, so I had to move ppx_have and discover out of `src`. There were a few updates needed to account for the changes to the language and the standard library. The switch to odoc also requires some changes to the documentation, and to move the default AST in ppx_have to a recent OCaml.
Prefixing module names with the name of the library is also deprecated, so ExtUnixAll, ExtUnixSpecific, and ExtUnixConfig only exist as ExtUnix.All, ExtUnix.Specific, and ExtUnix.Config now.
Dune now generate extunix.opam automatically, and the release process is done with dune-release which reads CHANGES.txt, tags the git repo, creates then publishes a tarball.

### CODE REVIEW

**Comment by MisterDA** [comment]
Hello @ygrek,
I believe that I have converged and that the PR is ready for review.
One thing that I haven't done is to port ppx_have to the newer ppxlib.

**Comment by ygrek** [comment]
(I didn't forget about this)

**Comment by MisterDA** [comment]
Yay! I think this is it.
The first batch of commits fixes compilation issues, the second batch removes oasis and migrates to dune. I've also ported the ad-hoc discover exe to dune-configurator. There's a workaround for a bug in configurator that's fixed now, but we have to wait for a new release of dune to remove the workaround. And finally, I've ported the ppx to ppxlib. It has a number of improvements over ocaml-migrate-parsetree.
Let me know what you think!

I also have some patches to make more functions work on Windows, but I intend to rebase them on this branch once it's merged.

File: appveyor.yml
Comment (kit-ty-kate) [inline appveyor.yml:7]:
```suggestion
  CYG_ROOT: C:\cygwin64
  OPAM_VERSION: 2.0.8
```
Could you try this? it might help the doubling of output in opam

File: dune-project
Comment (kit-ty-kate) [inline dune-project:9]: This most likely need a lower-bound version constraint, `>= "0.18.0"` at least, i suspect, due to the use of `Pconst_string` with 3 arguments.

File: appveyor.yml
Comment (MisterDA) [reply appveyor.yml:7]: I can reproduce in local with Opam 2.0.8, so this won't change it.

File: dune-project
Comment (MisterDA) [reply dune-project:9]: I want to use OCaml 4.12, so I'll go with 0.22.

File: appveyor.yml
Comment (kit-ty-kate) [reply appveyor.yml:7]: ah, I see. That sounds like an opam-on-windows™ bug indeed then :/ Have you tried with opam 2.1.0~beta4?

**Review by kit-ty-kate — APPROVED** [review]

**Comment by kit-ty-kate** [comment]
ping @ygrek if you have some time to review this

File: Makefile
Comment (ygrek) [inline Makefile:7]: I will keep old way of doing releases for now

**Review by ygrek — APPROVED** [review]
thanks, this looks neat

File: src/fexecve.c
Comment (MisterDA) [reply src/fexecve.c:19]: Because of this warning: a `const` string pointer is written into the non-const array.
```
dune build @install
         gcc src/fexecve.o
fexecve.c: In function ‘array_of_value’:
fexecve.c:19:12: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     arr[i] = String_val(Field(v, i));
            ^
```
However you're right, the copy isn't necessary and I believe a cast removing the const is enough to remove the warning, as the string isn't written into.

File: src/common.h
Comment (MisterDA) [reply src/common.h:4]: Sure, the end-result is the same, but we know these macros were introduced in 4.12.
I applied the same patch for the same issue here: https://github.com/AltGr/ocaml-mccs/pull/30.

**Comment by kit-ty-kate** [comment]
@ygrek is there a release planned by any chance? Having a release with this PR in would help us immensely switch away from OMP1 and/or camlp4 (given some projects use OMP2 + extunix and only extunix < 0.2.0 is available with this configuration)

**Comment by avsm** [comment]
@ygrek would you have time to cut a release with this PR included? Not having this is causing problems with the ocaml/opam base image builders (as they need newer ppxlib and extlib). Thanks!

**Comment by ygrek** [comment]
https://github.com/ocaml/opam-repository/pull/18478

**Pull request merged**
